Doc: `Headers` documentation (#817)
zio/zio-http: docs/website/docs/dsl/headers/index.md

# Headers

> ZIO-Http provide support for all HTTP (as defined in [RFC2616](https://datatracker.ietf.org/doc/html/rfc2616) ) headers and the library user's may define and use custom headers too.

Reviewer: ```suggestion
ZIO HTTP provides support for all HTTP (as defined in [RFC2616](https://datatracker.ietf.org/doc/html/rfc2616) ) headers and custom headers are supported too.
```

> ## Server side

Reviewer: ```suggestion
## Server-side
```

### Attaching Headers to `Response`

> In the server side implementation, `zio-http` is adding a collection of pre-defined headers to any response, according to http specification, but on top of them, user's of the library
> may add other headers, including custom headers.

Reviewer: ```suggestion
On the server-side, `ZIO-HTTP` is adding a collection of pre-defined headers to any response, according to the HTTP specification, but on top of them, users of the library may add other headers, including custom headers.
```

There are multiple ways to attach headers to a response:
- using `Response.addHeaders` API
- through `Response` constructors

> - using `Middlewares`

Reviewer: There are multiple ways to add headers in the response:
- Using `addHeaders` helper on response.
   ```scala
   val res = Response.ok.addHeader("X-MY-HEADER", "test")
   ```
- Through `Response` constructors.
   ```scala
   val res = Response(
          status = Status.OK,
          // Setting response header 
          headers = Headers.contentLength(0L),
          data = HttpData.empty
   ```
- Using `Middlewares`.
   ``` scala
   val app = Http.ok @@ Middleware.addHeader("X-MY-HEADER", "test")
   ```

Reviewer: This is skipped

        Response(
          status = Status.OK,
          // Setting response header 
          headers = Headers.contentLength(message.length.toLong), // adding CONTENT-LENGTH header
          data = HttpData.fromStream(ZStream.fromChunk(message)), // Encoding content using a ZStream
        )
      else { 
        // Adding a custom header to Response
        Response(status = Status.ACCEPTED, data = HttpData.fromChunk(message)).addHeader("X-MY-HEADER", "test")
      }
  }
}

```

The following example shows how Headers could be added to `Response` in the `Middleware` implementation: 

```scala

  /**
   * Creates an authentication middleware that only allows authenticated requests to be passed on to the app.
   */
  final def customAuth(
    verify: Headers => Boolean,
    responseHeaders: Headers = Headers.empty,
  ): HttpMiddleware[Any, Nothing] =
    Middleware.ifThenElse[Request](req => verify(req.getHeaders))(
      _ => Middleware.identity,
      _ => Middleware.fromHttp(Http.status(Status.FORBIDDEN).addHeaders(responseHeaders)),
    )

```

More examples:
- [BasicAuth](https://github.com/dream11/zio-http/blob/main/example/src/main/scala/BasicAuth.scala)

> - [Authentication](https://github.com/dream11/zio-http/blob/main/example/src/main/scala/Authentication.scala)

Reviewer: Make this section collapsible with Heading Detailed examples.

### Reading Headers from `Request`

> Server side implementation provides API that allows to read `Request` headers. The example above shows how also how the request headers could be read.
> In the example, the `ACCEPT` request header is read/checked to see if the `octet-streaming` response is acceptable for the client.
> 
> TBD: example

Reviewer: ```suggestion
On the Server-side you can read Request headers. The example below shows how the request headers could be read. In the example, the ACCEPT request header is read/checked to see if the octet-streaming response is acceptable for the client.
TBD:  add just request reading example here
```

> ## Client Side

Reviewer: ```suggestion
## Client-Side
```

### Adding headers to `Request` 

> ZIO-http provides a simple way to add headers to a client `Request`. Fo example, in most of the cases a client request should provide the `Host` header. 
> The sample below shows how a header could be added into a client request:

Reviewer: ```suggestion
ZIO-HTTP provides a simple way to add headers to a client Request. 
For example, in most cases, a client request should provide the Host header. The sample below shows how a header could be added to a client request:
```

```scala
import zhttp.http.{HeaderNames, HeaderValues, Headers}
import zhttp.service.{ChannelFactory, Client, EventLoopGroup}
import zio.{App, ExitCode, URIO, console}

object SimpleClientJson extends App {
  val env     = ChannelFactory.auto ++ EventLoopGroup.auto()
  val url     = "http://sports.api.decathlon.com/groups/water-aerobics"
  // Construct headers
  val headers = Headers.host("sports.api.decathlon.com").withAccept(HeaderValues.applicationJson) 

  val program = for {
    // Pass headers to request
    res  <- Client.request(url, headers)
    // List all response headers
    _    <- console.putStrLn(res.headers.toList.mkString("\n"))
    data <-
      // Check if response contains a specified header with a specified value.
      if (res.hasHeader(HeaderNames.contentType, HeaderValues.applicationJson))
        res.getBodyAsString
      else
        res.getBodyAsString

>     _    <- console.putStrLn { data }

Reviewer: just show case how to read headers
```scala
val responseHeaders: Task[Headers] =  Client.request(url).map(_.headers)
```

  } yield ()

  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = program.exitCode.provideCustomLayer(env)

}
```

> ### Reading headers from `Response`

Reviewer: you can move the above detailed example in the collapsible section here

TBD

## Headers DSL

> Headers DSL provides a ton of powerful operators that can be used to add, remove and modify headers.
> Same Headers API could be used on both client, server and middleware.

Reviewer: ```suggestion
Headers DSL provides plenty of powerful operators that can be used to add, remove and modify headers. Same Headers API could be used on both client, server, and middleware.
```

`zhttp.http.Headers`      - represents an immutable collection of headers i.e. essentially a `Chunk[(String, String)]`.

`zhttp.http.HeaderNames`  - commonly used header names.

`zhttp.http.HeaderValues` - commonly used header values

- Constructors:

> 

Reviewer: Add 1 line description here

```scala
import zhttp.http._

// create a simple Accept header:
val acceptHeader: Headers = Headers.accept(HeaderValues.applicationJson)

// create a basic authentication header:
val basicAuthHeader: Headers = Headers.basicAuthorizationHeader("username", "password")
```

> - Getters:

Reviewer: Add 1 line description here

```scala
import zhttp.http._

// retrieving the value of Accept header value:
val acceptHeader: Headers = Headers.accept(HeaderValues.applicationJson)
val acceptHeaderValue: Option[CharSequence] = acceptHeader.getAccept

// retrieving a bearer token from Authorization header:
val authorizationHeader: Headers                   = Headers.authorization("Bearer test")
val authorizationHeaderValue: Option[String]       = authorizationHeader.getBearerToken
```

> - Modifiers:

Reviewer: Add 1 line description here

```scala
import zhttp.http._

// add Accept header:
val headers = Headers.empty
val updatedHeadersList: Headers = headers.addHeaders(Headers.accept(HeaderValues.applicationJson))

// or if you prefer the builder pattern:

// add Host header:
val moreHeaders: Headers        = headers.withHost("zio-http.dream11.com")

```

> - Checks:

Reviewer: Add 1 line description here
